chat-android/src/main/java/com/ably/chat/RoomLifecycleManager.kt

```diff
@@ -1,119 +1,53 @@
+     * Implements code to free up underlying resources.
      * Spec: CHA-RL3h
      */
-    fun release()
-}
-
-internal abstract class ContributesToRoomLifecycleImpl(logger: Logger) : ContributesToRoomLifecycle {
-
-    private val discontinuityEmitter = DiscontinuityEmitter(logger)
-
-    override fun onDiscontinuity(listener: EmitsDiscontinuities.Listener): Subscription {
-        discontinuityEmitter.on(listener)
-        return Subscription {
-            discontinuityEmitter.off(listener)
-        }
-    }
-
-    override fun discontinuityDetected(reason: ErrorInfo?) {
-        discontinuityEmitter.emit("discontinuity", reason)
-    }
+    fun dispose()
 }
 
 /**
- * The order of precedence for lifecycle operations, passed to PriorityQueueExecutor which allows
- * us to ensure that internal operations take precedence over user-driven operations.
+ * The order of precedence for lifecycle operations, passed to PriorityQueueExecutor which
+ * allows us to ensure that certain operations are completed before others.
  */
 internal enum class LifecycleOperationPrecedence(val priority: Int) {
-    Internal(1),
-    Release(2),
-    AttachOrDetach(3),
+    Release(1),
+    AttachOrDetach(2),
 }
 
 /**
- * A map of contributors to pending discontinuity events.
- */
-internal typealias DiscontinuityEventMap = MutableMap<ContributesToRoomLifecycle, ErrorInfo?>
-
-/**
- * An internal interface that represents the result of a room attachment operation.
- */
-internal interface RoomAttachmentResult : NewRoomStatus {
-    val failedFeature: ContributesToRoomLifecycle?
-    val exception: AblyException
-}
-
-internal class DefaultRoomAttachmentResult : RoomAttachmentResult {
-    internal var statusField: RoomStatus = RoomStatus.Attached
-    override val status: RoomStatus
-        get() = statusField
-
-    internal var failedFeatureField: ContributesToRoomLifecycle? = null
-    override val failedFeature: ContributesToRoomLifecycle?
-        get() = failedFeatureField
-
-    internal var errorField: ErrorInfo? = null
-    override val error: ErrorInfo?
-        get() = errorField
-
-    internal var throwable: Throwable? = null
-
-    override val exception: AblyException
-        get() {
-            val errorInfo = errorField
-                ?: lifeCycleErrorInfo("unknown error in attach", ErrorCode.RoomLifecycleError)
-            return lifeCycleException(errorInfo, throwable)
-        }
-}
-
-/**
- * An implementation of the `Status` interface.
+ * An implementation of RoomLifecycleManager.
  * @internal
  */
 internal class RoomLifecycleManager(
+    private val room: DefaultRoom,
     private val roomScope: CoroutineScope,
-    private val statusLifecycle: DefaultRoomLifecycle,
-    private val contributors: List<ContributesToRoomLifecycle>,
+    private val statusManager: DefaultRoomStatusManager,
+    private val roomFeatures: List<RoomFeature>,
     roomLogger: Logger,
-) {
+) : DiscontinuityImpl(logger = roomLogger) {
```

Comment:
I would still prefer `Discontinuity by DiscontinuityImpl(logger = roomLogger)` don't like inheritance if it doesn't look natural and mostly done to extract part of business logic

Reply:
Actually I tried this, but then I can't access methods from `HandleDiscontinuity` interface :(

Reply:
OK, let's leave it than